File: official-plugins/desktop-ai-assistant/_ir/python/app/mcp/tools/echo.py
from typing import Any


def build_echo_tool(plugin_id: str) -> tuple[dict[str, Any], Any]:
    tool_definition = {
        "name": "assistant.python.echo",
        "description": "Echo payload from python sidecar for runtime diagnostics",
        "inputSchema": {
            "type": "object",
            "properties": {
                "message": {"type": "string"},
            },
            "required": [],
        },
    }

    async def tool_handler(arguments: dict[str, Any]) -> dict[str, Any]:
        message = str(arguments.get("message") or "ok")
        return {
            "message": message,
            "plugin_id": plugin_id,
            "service": "python-sidecar",
        }

    return tool_definition, tool_handler

File: official-plugins/desktop-ai-assistant/_ir/backend/src/mcp/tools/echo.ts
export interface McpToolDefinition {
  name: string;
  title: string;
  description: string;
  inputSchema: Record<string, unknown>;
}

export type McpToolHandler = (args: Record<string, unknown>) => Promise<Record<string, unknown>>;

export function buildEchoTool(pluginId: string): {
  tool: McpToolDefinition;
  handler: McpToolHandler;
} {
  const tool: McpToolDefinition = {
    name: "assistant.backend.echo",
    title: "Backend Echo",
    description: "Echo payload from Bun backend MCP",
    inputSchema: {
      type: "object",
      properties: {
        message: { type: "string" },
      },
      required: ["message"],
    },
  };

  const handler: McpToolHandler = async (args) => {
    const message = String(args.message ?? "").trim();
    return {
      content: [
        {
          type: "text",
          text: message || "empty",
        },
      ],
      meta: {
        plugin_id: pluginId,
        runtime: "bun",
      },
    };
  };

  return { tool, handler };
}
